README.md
# corrin
Automatically exported from code.google.com/p/corrin
这是一个服务器运维管理工具，主要特点：

管理具有SSH信任管理的运维拓扑服务器
只需在中央节点上进行安装和操作
依据需求进行节点筛选，批量执行
模块化架构设计
具有一定代码级别功能扩展能力
名字来源
corrin中文称为咕啉，取自维生素B12化学结构。

维生素B12是B族维生素中迄今为止发现最晚的一种。维生素B12是一种含有3价钴的多环系化合物，4个还原的吡咯环连在一起变成为1个咕啉大环，是维生素B12分子的核心。是红血球生成不可缺少的重要元素，如果严重缺乏，将导致恶性贫血。

服务器维护网络同咕啉环结构相似，希望成为运维工作的得力助手。

设计理念
从中央节点延展出2层、3层或者更多层的树状结构，以同心圆的方式展示，便成为星形散射装分布。

每台被操作服务器，即这个结构上的一个节点。这些节点信息，保存在中央机器的数据库中，由父子关系计算出行动路径，再依据一些配置执行命令。

日常运维工作中操作点比较多，但抽象出来无非是三种操作：执行命令、传输文件、记录信息。所以，实现这三种操作，便可以通过拆分操作然后进行排列组合即可实现功能上的扩展。

特点
利用数据关系，构建服务器节点网络
通过中央节点发送命令，控制节点操作，同时不限于中央对操作服务器操作。
实现任意节点间操作，如mysql数据库迁移，控制A节点备份，计算节点路径，传输，B节点恢复
架构设计
此工具使用Python语言编写

逻辑层	主要功能	Python模块	代码类
操作层	完成用户操作，主要命令行操作	cmd2	PizzaShell?
筛选层	实现网络节点条件筛选		Piece
逻辑层	实现单台目标或单功能操作		NodeNet?/Server/IPsec等等
　数据层	对数据库进行读写	sqlalchemy	t_server等以调用类名称定义的数据对象类
分发层	实现操作对象多线程分发		Parallel(待实现)
执行层	给目标服务器发送操作指令	Fabric	Server.execute
部署需求
Linux 5.+版本操作系统
python 2.7.3+
Fabric 1.8
sqlalchemy 0.8.1
cmd2 0.6.7
paramiko 1.12.0
pycrypto 2.6
mysqld 5.5+
prettytable 0.7.2
使用帮助
部署安装?
工具shell介绍
记录点滴


changes.md
#记录设计和开发过程的尝试和修改


# 设计修改 #
记录设计思想调整，使用模块调整

| 日期 | 内容 | 进度 |
|:-------|:-------|:-------|
| 2013-11-29 | 使用thread.start\_new\_thread的方式测试多线程执行失败 | 0% |
| 2013-11-30  | cmd命令，不再使用cmd2模块自带的参数解析，而是自主使用shlex和optparse配合使用，实现复杂命令用引号间隔正确执行的功能，此方法可以用在今后使用yaml配置命令的动态参数方法上 | 100% |
| 2013-11-29 | 重写cmd命令，使用shlex拆分命令行，再用optparse拆分参数，实现复杂命令输入 | 100% |
| 2013-11-24 | 开始学习twisted，计划使用此模块设计异步消息式操作方式 | 5% |
| 2013-12-05 | nagios命令，增加在centreon中创建监控主机和重启卫星功能。| 100% |
| 2013-12-09 | dbpizza.py 增加process\_list函数，用于统一处理节点或应用列表处理过程，但在传递函数参数时还存在问题，此功能不能使用；晚上参数问题无故解决。| 100% |
| 2013-12-10 |  增加配置文件config/pizza.ini，[pizza](pizza.md)节配置centerid，此数值对应t\_server表内的记录ID，用于在不同环境中构建不同的访问节点网络 | 100% |
| 2013-12-11 | 增加prettytable模块，将命令输出结果表格化显示 | 60% |
| 2013-12-11 | 问题：提示符增加颜色后，输入超过一行后，不会自动换行 2013-12-12 解决 详见issue | 100% |
| 2013-12-12 | 给NetNode和PizzaShell增加encoding属性，提供修改编码接口 | 100% |
| 2013-12-12 | 打印执行结果或运行结果，判断如果结果字符串是多行，采用交叉打印方式，如果只有一行，采用表格方式打印 | 100% |
| 2013-12-13 | 解决数据库连接超时问题，详见Issue | 100% |
| 2013-12-13 | 学习Tornado，可以利用写网络服务和web服务 http://doc.outofmemory.cn/python/tornado/#_9 | 3% |
| 2013-12-17 | 修正了主程序中的多线程代码，"cmd"命令现可以以多线程方式执行命令，但还有1个问题：多线程时消耗时间计算不正确 | 80% |
| 2013-12-17 | 试用了saltstack，突然有种另造轮子的感觉 | 100% |
| 2013-12-18 | 将前身dbpizza的代码迁移到此，修改了对应的主文件名为corrin.py | 100% |
| 2013-12-18 | 发现使用Feature时，那个节点信息不能完全加载。是因为Feature是纵向添加，Server是横向添加。在breed时只判断是否有子节点，如果有就不加载：修改纵向加载方式，每个节点加载全部子节点 | 100%|
| 2013-12-19 | 开发 Iptables 功能，预期能够实现 Iptables 的批量、安全、入库管理，另外能够实现智能识别端口开放情况并以可读样式列出 | 10% |
| 2013-12-19 | Open Sourcing a Python Project the Right Way  http://www.jeffknupp.com/blog/2013/08/16/open-sourcing-a-python-project-the-right-way/ | 0% |
| 2013-12-24 | 增加备份节点功能，在生成秘钥、防火墙时，要读取备份节点信息，保障互备可用 | 0% |
| 2014-01-02 | 增加Operation类，作为操作的基类，收纳初始化dbsession和通用获取数据方法 | 100% |
| 2014-01-07 | 为 dbi 增加自动创建不存在表的功能，详细书写了代码中的表结构定义，保证额外属性没有被遗漏（如自动更新时间戳等）；另外目前还不能自动 ALTER TABLE | 50% |
| 2014-01-20 | 将数据操作移到dbi数据表基类中 | 100% |
|2014-01-21| 使用新数据读取接口，重构node.py中类调用，除iptables以外 | 100% |
| 2014-01-24 | 尝试使用multiprocessing写多进程处理，效果不错，整体结构失败中 | 10% |
| 2014-01-25 | Happy Chinese New Year | 100% |
| 2014-03-24 | 当用户要进行一项多步骤；子步骤需要可重复；调用过程要先写下来，复查，再执行时；用户需要可以固化的脚本，而不是交互式命令行 | 0% |

dbi.py
#!/usr/bin/env python
#coding:utf-8
# Author:  Haiyang Peng
# Purpose:
# Created: 2013/6/28
import sqlalchemy
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, Integer, BigInteger, VARCHAR, Text, DateTime, DATETIME, SMALLINT, TIMESTAMP, Boolean
from sqlalchemy.dialects.mysql import TINYINT
from sqlalchemy.types import SchemaType, TypeDecorator, Enum
from sqlalchemy.orm import sessionmaker
from sqlalchemy.sql.expression import text
from sqlalchemy.orm import relationship,backref
import ConfigParser
import time, datetime
import sys, os.path
from alembic.config import Config
from alembic import command


def _get_db_string():
    config = ConfigParser.SafeConfigParser()
    base_path = os.path.dirname(os.path.abspath(__file__))
    config.read(os.path.join(base_path, "config/dbi.ini"))
    dblink = """%s://%s%s@%s/%s?charset=%s""" % (config.get('database', 'engine'),
                                                 config.get('database', 'user'),
                                                 config.get('database', 'password') if config.get('database',
                                                                                                  'password') else '',
                                                 config.get('database', 'host'),
                                                 config.get('database', 'database'),
                                                 config.get('database', 'charset'))
    return dblink


engine = create_engine(_get_db_string(), pool_size=4, pool_recycle=10, max_overflow=10)

Base = declarative_base()
Session = sessionmaker(bind=engine)
#session = Session()



#这样写会不会每次调用模块都会被调用，相同的功能添加到init_db()中了
#Base.metadata.create_all(engine)

def init_db():
    BaseModel.metadata.create_all(engine)
    base_path = os.path.dirname(os.path.abspath(__file__))
    alembic_cfg_file = os.path.join(base_path, "alembic.ini")
    alembic_cfg = Config(alembic_cfg_file)
    ## Initialize the alembic version after fully create a new database by "create_all"
    # command.stamp(alembic_cfg, "head")
def drop_db():
    BaseModel.metadata.drop_all(engine)

class BaseModel(Base):
    __abstract__ = True
    __table_args__ = {
        'mysql_engine':'InnoDB',
        'mysql_charset':'utf8'
        }

    # use a session every record
    # session=Session()

    @classmethod
    def get(cls,  value ,filter='id', columns=None, lock_mode=None,first=False):
        result=None
        try:
            session=Session()
            if hasattr(cls,filter):
                scalar = False
                if columns:
                    if isinstance(columns, (tuple, list)):
                        query = session.query(*columns)
                    else:
                        scalar = True
                        query = session.query(columns)
                else:
                    query = session.query(cls)
                if lock_mode:
                    query = query.with_lockmode(lock_mode)
                if isinstance(value,(list,tuple)):
                    query=query.filter(getattr(cls,filter).in_(value))
                else:
                    query=query.filter(getattr(cls,filter) == value)
                if scalar:
                    result = query.scalar()
                if first:
                    result = query.first()
                else:
                    result = query.all()
            else:
                result = None
        except Exception,e:
            print e
            if session:
                session.rollback()
            result = None
        finally:
            if session:
                session.close()
            return result


    #@classmethod
    #def count_all(cls, lock_mode=None):
        #query = cls.session.query(func.count('*')).select_from(cls)
        #if lock_mode:
            #query = query.with_lockmode(lock_mode)
        #return query.scalar()


    #@classmethod
    #def exist(cls,  id, lock_mode=None):
        #if hasattr(cls, 'id'):
            #query = cls.session.query(func.count('*')).select_from(cls).filter(cls.id == id)
            #if lock_mode:
                #query = query.with_lockmode(lock_mode)
            #return query.scalar() > 0
        #return False




    @classmethod
    def set_attrs(cls,  id, **attrs):
        try:
            session=Session()
            if hasattr(cls, 'id'):
                session.query(cls).filter(cls.id == id).update(attrs)
                session.commit()
        except Exception,e:
            print e
            if session:
                session.rollback()
        finally:
            if session:
                session.close()


    @classmethod
    def add(cls, **attrs):
        try:
            session=Session()
            cls.session.add(cls(**attrs))
            cls.session.commit()
        except Exception,e:
            print e
            if session:
                session.rollback()
        finally:
            if session:
                session.close()

    @classmethod
    def delete(cls,value,filter='id'):
        try:
            session=Session()
            for ins in cls.get(value,filter):
                session.delete(ins)
            session.commit()
        except Exception,e:
            print e
            if session:
                session.rollback()
        finally:
            if session:
                session.close()





class t_region(BaseModel):
    __tablename__ = 't_region'
    #MySQL 5.5 开始支持存储 4 字节的 UTF-8 编码的字符了，iOS 里自带的 emoji（如 🍎 字符）就属于这种。
    #如果是对表来设置的话，可以把上面代码中的 utf8 改成 utf8mb4，DB_CONNECT_STRING 里的 charset 也这样更改。
    id = Column(Integer, primary_key=True)
    code = Column(VARCHAR(3))
    name = Column(VARCHAR(50))


class t_product(BaseModel):
    __tablename__ = 't_product'
    id = Column(Integer, primary_key=True)
    code = Column(VARCHAR(20))
    name = Column(VARCHAR(50))


class t_server(BaseModel):
    __tablename__ = 't_server'
    __searchword__ = None
    id = Column(Integer, primary_key=True)
    pid = Column(Integer, nullable=True)
    # 记录上层硬件所属ID，用途：虚拟机-》物理机
    mid = Column(Integer, nullable=True)
    region = Column(Enum('hk', 'vn', 'id', 'in', 'tw', 'th', 'us', 'my', 'cn'))
    product = Column(
        Enum('tlbb', 'ldj', 'taoyuan', 'guyu', 'totem', 'specialforce', 'gamefuse', 'oppaplay', 'gamiction', 'cuaban',
             'davinci', 'swordgirls', 'zszw', 'common', 'pengyou'))
    role = Column(Enum('cc', 'backup', 'db'))
    loginuser = Column(VARCHAR(40), server_default='root')
    description = Column(VARCHAR(250))
    ip_oper = Column(VARCHAR(16))
    ip_private = Column(VARCHAR(16))
    ip_public = Column(VARCHAR(16))
    ip_ilo = Column(VARCHAR(16))
    is_reserve = Column(Boolean, server_default='0')
    dbms = Column(Enum('MySQL', 'Oracle', 'MSSQL'))
    vender = Column(Enum('Dell', 'HP', 'VMware', 'Intel', 'Xen'))
    model = Column(VARCHAR(100))
    os_type = Column(Enum('Linux', 'Windows'))
    os_release = Column(Enum('RHEL_5_3', 'RHEL_4_8', 'RHEL_4_6', 'CENT_6_3', 'WIN2003', 'WIN2008'))
    os_arch = Column(Enum('x86_64', 'i386'))
    ip_monitor = Column(VARCHAR(16))
    ip_ntp_server = Column(VARCHAR(16))
    serial = Column(VARCHAR(50))
    is_online = Column(Boolean, server_default='0')
    update_time = Column(TIMESTAMP, server_default=text('0 ON UPDATE CURRENT_TIMESTAMP'))
    create_time = Column(TIMESTAMP, server_default=text('0'))
    is_deleted = Column(Boolean)

    def __repr__(self):
        return "<Server('%s','%s','%s')>" % (self.region, self.product, self.ip_oper)

    @classmethod
    def _all_words(cls):
        if cls.__searchword__:
            return cls.__searchword__
        words = []
        words += cls._get_word(t_server.region)
        words += cls._get_word(t_server.product)
        words += cls._get_word(t_server.role)
        words += cls._get_word(t_server.dbms)
        words += cls._get_word(t_server.vender)
        words += cls._get_word(t_server.os_type)
        words += cls._get_word(t_server.os_release)
        words += cls._get_word(t_server.os_arch)
        words += [('reserve', t_server.is_reserve)]
        words = dict(words)
        cls.__searchword__ = words
        return cls.__searchword__

    @classmethod
    def _get_word(cls, col):
        try:
            session=Session()
            ret = session.query(col).filter(col != None).distinct().all()
        finally:
            if session:
                session.close()
        return [(str(x[0]), col) for x in ret]

    @classmethod
    def piece(cls, keywords):
        import string

        keywords = string.split(keywords, ',')
        words = cls._all_words()
        knife = {}
        for i in keywords:
            if i in words.keys():
                knife[i] = words[i]
        if len(knife.keys()) <= 0:
            return []
        try:
            session=Session()
            result = session.query(t_server)
        finally:
            if session:
                session.close()
        for value, col in knife.iteritems():
            result = result.filter(col == value)
        result = result.all()
        return [i.id for i in result]


class t_feature(BaseModel):
    __tablename__ = 't_feature'
    id = Column(Integer, primary_key=True)
    pid = Column(Integer)
    feature = Column(VARCHAR(50))
    detail = Column(VARCHAR(50))
    server_id = Column(Integer)


class t_ipsec(BaseModel):
    __tablename__ = 't_ipsec'
    id = Column(Integer, primary_key=True)
    server_id = Column(Integer, index=True)
    chain = Column(Enum('INPUT', 'OUTPUT', 'FORWARD'))
    source_addr = Column(VARCHAR(50))
    dest_addr = Column(VARCHAR(50))
    protocal = Column(Enum('tcp', 'udp', 'icmp', 'all'))
    dport = Column(VARCHAR(50))
    #
    status = Column(Integer)
    description = Column(VARCHAR(100))
    create_time = Column(DATETIME)
    modify_time = Column(DATETIME)

    def __init__(self, server_id, protocal, source_addr, dport, description, status=0, chain='INPUT'):
        self.chain = chain
        self.server_id = server_id
        self.chain = chain
        self.source_addr = source_addr
        self.protocal = protocal
        self.dport = dport
        self.status = status
        self.description = description
        self.create_time = datetime.datetime.now()
        self.modify_time = datetime.datetime.now()


class t_sysinfo(BaseModel):
    __tablename__ = 't_sysinfo'
    id = Column(Integer, primary_key=True)
    need_id = Column(Integer)
    need_value = Column(VARCHAR(50))
    check_name = Column(VARCHAR(40))
    check_cmd = Column(VARCHAR(255))
    sys_type = Column(Enum('Windows', 'Linux', 'All'), index=True)
    result_reg = Column(VARCHAR(50))
    record_table = Column(VARCHAR(50))
    record_field = Column(VARCHAR(50))


class t_crontab(BaseModel):
    __tablename__ = 't_crontab'
    id = Column(Integer, primary_key=True)
    server_id = Column(Integer)
    pminute = Column(VARCHAR(20))
    phour = Column(VARCHAR(20))
    pday = Column(VARCHAR(20))
    pmonth = Column(VARCHAR(20))
    pweek = Column(VARCHAR(20))
    process = Column(VARCHAR(400))
    status = Column(SMALLINT)
    user = Column(VARCHAR(30))
    group = Column(VARCHAR(20))
    description = Column(VARCHAR(100))
    operator = Column(VARCHAR(30))
    create_time = Column(TIMESTAMP)
    update_time = Column(TIMESTAMP)


class t_iptables(BaseModel):
    __tablename__ = 't_iptables'
    id = Column(Integer, primary_key=True)
    server_id = Column(Integer)
    trx_id = Column(VARCHAR(16))
    trx_time = Column(TIMESTAMP, server_default=text('CURRENT_TIMESTAMP'))


class t_iptables_rules(BaseModel):
    __tablename__ = 't_iptables_rules'
    id = Column(Integer, primary_key=True)
    trx_id = Column(VARCHAR(16))
    index = Column(TINYINT)
    table = Column(VARCHAR(16))
    chain = Column(VARCHAR(50))
    opt = Column(VARCHAR(150))
    arg = Column(VARCHAR(150))


